From 096d649df2f4d865cbaece1f5bc189511b9d70d8 Mon Sep 17 00:00:00 2001
From: Brahma Reddy Battula <brahma@apache.org>
Date: Mon, 12 Feb 2018 22:14:34 +0530
Subject: [PATCH 2772/2848] HDFS-8693. Addendum patch to execute the command
 using UGI. Contributed by Brahma Reddy Battula.

(cherry picked from commit 35c17351cab645dcc72e0d2ae1608507aa787ffb)
(cherry picked from commit 24af0bc144b7e89fc2ea8d2f49b2c8eeb5a2b725)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockPoolManager.java

Change-Id: I8561572274d0a458da91464814c5c69da84cb272
---
 .../hdfs/server/datanode/BlockPoolManager.java     |   19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockPoolManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockPoolManager.java
index 9f38995..45807e8 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockPoolManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockPoolManager.java
@@ -229,10 +229,23 @@ private void doRefreshNamenodes(
           Joiner.on(",").useForNull("<default>").join(toRefresh));
       
       for (String nsToRefresh : toRefresh) {
-        BPOfferService bpos = bpByNameserviceId.get(nsToRefresh);
-        ArrayList<InetSocketAddress> addrs =
+        final BPOfferService bpos = bpByNameserviceId.get(nsToRefresh);
+        final ArrayList<InetSocketAddress> addrs =
           Lists.newArrayList(addrMap.get(nsToRefresh).values());
-        bpos.refreshNNList(addrs);
+        try {
+          UserGroupInformation.getLoginUser()
+              .doAs(new PrivilegedExceptionAction<Object>() {
+                @Override
+                public Object run() throws Exception {
+                  bpos.refreshNNList(addrs);
+                  return null;
+                }
+              });
+        } catch (InterruptedException ex) {
+          IOException ioe = new IOException();
+          ioe.initCause(ex.getCause());
+          throw ioe;
+        }
       }
     }
   }
-- 
1.7.9.5

